File: py/syntax/exception/exception__.py
#!/usr/bin/env python
# _*_  coding: utf-8 _*_

#*********************************************************************
# Author        : cuiyunpeng
# Email         : @163.com
# Create time   : 2020-07-24 05:52
# Last modified : 2020-07-24 05:52
# Filename      : exception__.py
# Description   :
#*********************************************************************
def this_fail():
    return 5/0

if __name__ ==  "__main__":
    # while True:
    #     try:
    #         x = int(input("Enter a intger: "))
    #         break
    #     except ValueError:
    #         print("The interger is wrong, enter agin: ")
    #
    # x = 1
    # if x > 5:
        # raise Exception("{} beyond 5".format(x))
    # try:
    #     this_fail()
    # except ZeroDivisionError as err:
    #     print("The err is {}".format(err))
    # finally:
    #     print("handel done.")
    #
    try:
        raise NameError("hellow")
    except NameError:
        print("The err is ")
        raise

File: cpp/syntax/exception/exception.cpp
/*********************************************************************
 * Author        : cuiyunpeng
 * Email         : @163.com
 * Create time   : 2020-05-16 21:40
 * Last modified : 2020-05-16 21:40
 * Filename      : exception.cpp
 * Description   : 
*********************************************************************/

#include <iostream>
#include <exception>

using namespace std;

#define CUSTOM_EXCEPTION
double division(int a, int b)
{
    if (b == 0) {
        throw "Division by zero condition!";
    }
    return a/b;
}
#if defined ZERO_DIVISION
int main(int argc, char **argv)
{
    int x = 50; 
    int y = 0;
    double z = 0;
    
    try {
        z = division(x, y);
        cout << z << endl; 

    } catch(const char *msg) {
        cerr << msg << endl;
    }
    return 0;
}
#elif defined  CUSTOM_EXCEPTION
struct exception_s {
    const char *what() const throw() {
        return "C++ exception."
    }
}

int main(int argc, char **argv)
{
    int x = 50; 
    int y = 0;
    double z = 0;
    
    try {
        throw my_exception_s();
    } catch (exception_s &e){
        cerr << e.what() << endl;
    }
    return 0;
}
#else

#endif

